google/cloud/storage/_signing.py

```diff
@@ -666,3 +665,33 @@ def _sign_message(message, access_token, service_account_email):
 
     data = json.loads(response.data.decode("utf-8"))
     return data["signature"]
+
+
+def _url_encode(query_params):
+    """Encode query params into URL.
+
+    :type query_params: dict
+    :param query_params: Query params to be encoded.
+
+    :rtype: str
+    :returns: URL encoded query params.
+    """
+    params = []
+    for name, value in query_params.items():
+        params.append(_quote_param(name) + "=" + _quote_param(value))
+
+    return "&".join(sorted(params))
+
+
+def _quote_param(param):
+    """Quote query param.
+
+    :type param: dict
+    :param param: Query param to be encoded.
+
+    :rtype: str
+    :returns: URL encoded query param.
+    """
+    if not isinstance(param, bytes):
+        param = str(param)
```

Comment:
We cannot use `str` here:  it returns bytes only under Python2!  Instead, we can let `quote` do the conversions for us:  it uses `utf-8` by default to encode non-ASCII characters.

Reply:
Not sure I understand. I've taken this conversion from original `urlencode()` (I mean, I've tried not to diverge too much from it). If we'll erase these lines, we'll get a lot of `TypeError` from conformance tests. `quote()` [seems to be expecting](https://github.com/python/cpython/blob/8dbdf5f275c6462bb522bcf3a29054239d72989d/Lib/urllib/parse.py#L806-L818) only `str` or `bytes` as an arg

Reply:
The example you link here doesn't use `str()` to convert anything: t uses `string.encode` to convert *text* to bytes, then (in `quote_from_bytes`) does the quoting at the bytes level, and finally returns *text* using `''.join()`.

I'm not sure what you mean about a `TypeError`:  do the conformance tests pass anything other than text / bytes as a value?

Reply:
Yes, `X-Goog-Expires` parameter comes in as an `int`, so all of the conformance tests are failing, as all of them have:
https://github.com/googleapis/python-storage/blob/23df542d0669852b05139023d5ef1ae14a09f4c7/tests/unit/url_signer_v4_test_data.json#L7-L7

as well as `test_with_access_token`, `test_w_response_type`, `test_w_response_disposition`, etc. - all with the same `TypeError`, because of expiration parameter.

> The example you link here doesn't use str()

No, no, I've linked `quote()` itself to see that it's expecting `str` or `bytes`. If we're speaking about conversion (which I'm keeping), than we should look [here](https://github.com/python/cpython/blob/0d860dd43c72dc7046a5d18fc72d495cadd4a2df/Lib/urllib/parse.py#L899-L907). Thus, if `k` or `v` is not `bytes`, they're converting it into `str`.

It's the `urllib.parse.urlencode()` behavior, so I suppose this conversion was always here

Reply:
If this `_quote_param` function is meant to take only `str` or `bytes` (as you've updated the docstring to indicate), then it is up to the caller to handle converting any other types:  this method should *not* coerce using `str`.

Reply:
No, it actually will deal with any type. Fixed

Reply:
As written this seems to handle python 2 and python 3. I am going to leave this as is. I agree the coercion isn't ideal.